File: Input Sanitizer/cpp/xss_filter.cpp
#include <iostream>
#include <string>
#include <regex>
#include <vector>

/**
 * V2.0: Enhanced Sanitization Strategy (Defense in Depth).
 * 
 * 1. Tag Stripping: Removes dangerous tags like <script>, <iframe>.
 * 2. Attribute Stripping: Removes dangerous attributes like 'onload', 'onclick'.
 * 3. Protocol Stripping: Removes 'javascript:' URIs.
 */
std::string sanitize_input_v2(std::string input) {
    // 1. Tag Stripping (Existing logic)
    std::regex tag_pattern(R"(<(script|iframe|object|embed)\b[^>]*>[\s\S]*?</\1>)", 
                           std::regex_constants::ECMAScript | std::regex_constants::icase);
    input = std::regex_replace(input, tag_pattern, "");

    // 2. Event Handler Stripping (e.g., onload=..., onerror=...)
    // Looks for ' on<event>=' pattern.
    std::regex event_pattern(R"(\s+on\w+\s*=)", 
                             std::regex_constants::ECMAScript | std::regex_constants::icase);
    input = std::regex_replace(input, event_pattern, " data-unsafe-attr=");

    // 3. Javascript Protocol Stripping (e.g., href="javascript:")
    std::regex js_proto_pattern(R"(javascript:)", 
                                std::regex_constants::ECMAScript | std::regex_constants::icase);
    input = std::regex_replace(input, js_proto_pattern, "unsafe:");

    return input;
}

/**
 * Output Encoding: The gold standard for XSS prevention.
 * Converts special characters into their HTML Entity equivalents.
 */
std::string html_encode(const std::string& data) {
    std::string buffer;
    buffer.reserve(data.size());
    for(size_t pos = 0; pos != data.size(); ++pos) {
        switch(data[pos]) {
            case '&':  buffer.append("&amp;");       break;
            case '"': buffer.append("&quot;");      break;
            case '\'': buffer.append("&apos;");      break;
            case '<':  buffer.append("&lt;");        break;
            case '>':  buffer.append("&gt;");        break;
            default:   buffer.append(&data[pos], 1); break;
        }
    }
    return buffer;
}

int main() {
    std::string dirty_string = 
        "Hello. <script>alert('XSS')</script>\n"
        "<b>Bold is safe.</b>\n"
        "<a href='javascript:alert(1)'>Dangerous Link</a>\n"
        "<img src=x onerror=alert('Event Handler XSS')>";

    std::cout << "Original Input:" << std::endl;
    std::cout << "----------------------------------------" << std::endl;
    std::cout << dirty_string << std::endl;
    std::cout << "----------------------------------------" << std::endl;

    // Approach 1: Enhanced Sanitization
    std::string clean_string = sanitize_input_v2(dirty_string);
    std::cout << "\n[Approach 1] Enhanced Sanitization (Filter Bad):" << std::endl;
    std::cout << "----------------------------------------" << std::endl;
    std::cout << clean_string << std::endl;
    std::cout << "----------------------------------------" << std::endl;

    // Approach 2: Output Encoding
    std::string encoded_string = html_encode(dirty_string);
    std::cout << "\n[Approach 2] Output Encoding (Safest for Text):" << std::endl;
    std::cout << "----------------------------------------" << std::endl;
    std::cout << encoded_string << std::endl;
    std::cout << "----------------------------------------" << std::endl;

    return 0;
}

File: Input Sanitizer/Python/xss_filter.py
import re
from html.parser import HTMLParser
import html

class WhitelistFilter(HTMLParser):
    def __init__(self):
        super().__init__()
        self.reset()
        self.sanitized_data = []
        # Whitelist: Only these tags are allowed.
        self.allowed_tags = {'b', 'i', 'u', 'p', 'br', 'a', 'strong', 'em'}
        # Whitelist: Only these attributes are allowed on specific tags.
        self.allowed_attrs = {
            'a': {'href', 'title'},
            'img': {'src', 'alt'} # Note: img not in allowed_tags above, just example if added
        }

    def handle_starttag(self, tag, attrs):
        if tag in self.allowed_tags:
            clean_attrs = []
            for attr, value in attrs:
                # 1. Attribute Whitelisting
                if tag in self.allowed_attrs and attr in self.allowed_attrs[tag]:
                    # 2. Protocol Validation (No javascript:)
                    if attr in ['href', 'src']:
                        if value.strip().lower().startswith('javascript:'):
                            continue # Skip dangerous protocol
                    clean_attrs.append(f'{attr}="{html.escape(value)}"')
            
            # Reconstruct the tag
            if clean_attrs:
                self.sanitized_data.append(f'<{tag} {" ".join(clean_attrs)}>')
            else:
                self.sanitized_data.append(f'<{tag}>')

    def handle_endtag(self, tag):
        if tag in self.allowed_tags:
            self.sanitized_data.append(f'</{tag}>')

    def handle_data(self, data):
        # 3. Output Encoding for text content
        self.sanitized_data.append(html.escape(data))

    def get_clean_data(self):
        return "".join(self.sanitized_data)

def sanitize_input_v2(user_input):
    """
    V2.0: Uses a Whitelist approach via HTMLParser.
    Only allows specific safe tags. Encodes everything else.
    """
    parser = WhitelistFilter()
    parser.feed(user_input)
    return parser.get_clean_data()

def encode_output(user_input):
    """
    Strategy: Output Encoding.
    Converts special characters to HTML entities.
    Safest general approach if you don't need HTML formatting.
    """
    return html.escape(user_input)

def main():
    dirty_string = (
        "Hello. <script>alert('XSS')</script>\n"
        "<b>Bold Text is fine.</b>\n"
        "<a href='javascript:alert(1)'>Dangerous Link</a>\n"
        "<img src=x onerror=alert('Event Handler XSS')>\n"
        "<p onclick='stealCookies()'>Click me</p>"
    )

    print("Original Input:")
    print("-" * 40)
    print(dirty_string)
    print("-" * 40)

    # Approach 1: Whitelist Sanitization (Allow safe HTML, strip bad)
    clean_html = sanitize_input_v2(dirty_string)
    print("\n[Approach 1] Whitelist Sanitization (Allow safe HTML):")
    print("-" * 40)
    print(clean_html)
    print("-" * 40)

    # Approach 2: Output Encoding (Treat everything as text)
    encoded_text = encode_output(dirty_string)
    print("\n[Approach 2] Output Encoding (Safest for plain text):")
    print("-" * 40)
    print(encoded_text)
    print("-" * 40)

if __name__ == "__main__":
    main()
